ssi-ldp/src/lib.rs

```diff
@@ -409,25 +418,279 @@ async fn to_jws_payload(
+    challenge_bytes.extend_from_slice(nonce_bytes.as_slice());
+    let challenge = ProofChallenge::hash(&challenge_bytes);
+
+    let bbs_proof = Prover::generate_signature_pok(pok, &challenge).unwrap();
+    let bbs_proof_bytes = bbs_proof.to_bytes_compressed_form();
+    let bbs_proof_str = base64::encode_config(bbs_proof_bytes.as_slice(), base64::URL_SAFE_NO_PAD);
+    let proof_str = header_str.clone() + ".." + bbs_proof_str.as_str();
+
+    let mut proof_with_new_sig = proof.clone();
+    proof_with_new_sig.jws = Some(proof_str);  // todo: change to proof/proofValue
+    Ok(proof_with_new_sig)
+}
+
+fn rename_blank_node_labels(orig: &Vec<String>) -> Vec<String> {
+    use std::collections::HashMap;
+
+    // hash maps for properties, parents, and substitutions
+    let mut blank_node_props: HashMap<String, String> = HashMap::new();
+    let mut blank_node_parents: HashMap<String, String> = HashMap::new();
+    let mut blank_node_subs: HashMap<String, String> = HashMap::new();
+
+    let mut root_node: Option<&str> = None;
+
+    for nq in orig.iter() {
+        let split: Vec<&str> = nq.split(" ").collect();
+        let left = split[0];
+        let middle = split[1];
+        let right = split[2];
+
+        if middle.ends_with("/credentials#credentialSubject>") {
+            root_node = Some(left);
+        }
+    }
+
+    blank_node_subs.insert(root_node.unwrap().to_owned(), "_d:0".to_owned());
+
+    for nq in orig.iter() {
+        let split: Vec<&str> = nq.split(" ").collect();
+        let left = split[0];
+        let middle = split[1];
+        let right = split[2];
+        if right.starts_with("_:c") {
+            eprintln!("property={}, blank node label={}", middle, right);
+            blank_node_props.insert(right.to_owned(), middle.to_owned());
+            blank_node_parents.insert(right.to_owned(), left.to_owned());
+        }
+    }
+
+    //eprintln!("properties map: {:?}", &blank_node_props);
+    //eprintln!("parents map: {:?}", &blank_node_parents);
+    // at this point, we have properties and parents for each blank node label
+
+    // iterate through the n-quads and construct the substitutions
+    // key is n-quad, value is parent, either another n-quad or an identifier
+    for (key, value) in blank_node_parents.iter() {
+        //eprintln!("key: {}, value: {}", &key, &value);
+
+        let mut path = String::from("");
+        let mut curr = key.as_str();
+        while curr != root_node.unwrap() {
+            let prop = blank_node_props[curr].as_str();
+            path.push_str(prop);
+            path.push(':');
+
+            let parent = blank_node_parents[curr].as_str();
+            if parent.starts_with("_:c") {
+                curr = blank_node_parents[curr].as_str();
+            } else {
+                path.push_str(parent);
+                path.push(':');
+                break;
+            }
+        }
+
+        let hash = sha256(path.as_bytes());
```

Comment:
It is worth investigating whether using SHA-256 is a good idea. Perhaps this choice of hash function can be attacked.